Parallel builders

iurly opened the pull request.

iurly committed:
parallel builds: mount driver code readonly and run cmake for ebpf

this will be the preliminary step to parallel builds:

====
- part 1: mount code readonly
====
- part 2: fix handling of bpf probes when source is mounted readonly

the previous implementation, with the driver folder mounted in r/w,
was working through some side effects of building the kmod:
it would create files (e.g. driver_config.h) which would be
available to the subsequent invocation for eBPF.
Now that such folders are no longer shared through the host filesystem,
they would get the artifacts from whatever build was done on the host
prior to introducing the parallel builders.
The end result was that we would get probes built with the wrong
version information.

So fix this by having an explicit run of cmake on it for the
eBPF case, too.
Also, it would be a good idea to run the builder off a clean
git checkout for the driver.

iurly committed:
parallel builders: do not name containers

if we run builders in parallel we can't have named containers

iurly committed:
execute builders in parallel (using affinity)

after making all the necessary fixes for parallelism
learned the hard way, finally enable parallel execution,
using affinity as a measure of the available CPUs.

Also, add some feedback about the stuff run in the futures.

iurly committed:
parallel builders: introduce locking for docker build

if we introduce parallel builders, we need to make sure that
only one 'docker build' command is executed at any given time,
otherwise we might get weird results.

iurly committed:
debug and print timing

gnosek commented on probe_builder/__init__.py:
```diff
@@ -96,7 +97,6 @@ def get_kernels(self, _workspace, packages, _download_config, filter):
 def cli(debug):
     init_logging(debug)
 
-
```
If you touch this file again, maybe remove this change :)

iurly replied:
done

gnosek commented on probe_builder/__init__.py:
```diff
@@ -124,8 +124,16 @@ def build(builder_image_prefix,
+            future = executor.submit(distro_builder.build_kernel, workspace, probe, distro.builder_distro, release, target)
+            kernels_futures.append((release, future))
+
+    for release, future in kernels_futures:
```
I know ~nothing about python futures but I assume this waits for the futures to be resolved, right?

deepskyblue86 replied:
@gnosek `result()` is the thing https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result

gnosek commented on probe_builder/__init__.py:
```diff
@@ -124,8 +124,16 @@ def build(builder_image_prefix,
+            kernels_futures.append((release, future))
+
+    for release, future in kernels_futures:
+        print("Release: {}, Result: {}".format(release, future.result()))
```
Do we want to use the logger here?

iurly replied:
good question, I didn't give it too much thought. I guess I went straight for a print() because that's essentially a final report, so there would be no added value in having a timestamp in each line.
At the end of the day we might even want to produce some structured artifact to be archived, such as an XML or HTML file. What do you think?

gnosek commented on probe_builder/builder/choose_builder.py:
```diff
@@ -146,6 +148,7 @@ def choose_gcc_dockerfile(builder_source, builder_distro, kernel_dir):
     dockerfile, _, support_bpf = dockerfile_version
     # get a tag we'll use for builders
     tag = dockerfile.replace('Dockerfile.', '')
+    logger.debug('elected dockerfile: {}'.format(dockerfile))
```
chosen? selected?

gnosek replied:
bump

iurly replied:
ok done u1

gnosek commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -88,14 +91,21 @@ def build_kernel(self, workspace, probe, builder_distro, release, target):
         builder_image.build(workspace, dockerfile, dockerfile_tag)
+        ts1 = time.time()
+
+        logger.info("Docker building of {} took {:.2f}s".format(dockerfile, ts1-ts0))
+        took = ts1 - ts0
```
```suggestion
        took = ts1 - ts0
        logger.info("Docker building of {} took {:.2f}s".format(dockerfile, took))
```

iurly replied:
thanks! I actually also caught a serious error while looking this up!

iurly replied:
Changes accepted with the latest rework (from ~1 month ago)

gnosek commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -54,15 +55,17 @@ def build_kernel_impl(cls, config_hash, container_name, image_name, kernel_dir,
             return
 
-        docker.rm(container_name)
+        #docker.rm(container_name)
```
Maybe just delete it?

gnosek commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -88,14 +91,21 @@ def build_kernel(self, workspace, probe, builder_distro, release, target):
 
         image_name = '{}sysdig-probe-builder:{}'.format(workspace.image_prefix, dockerfile_tag)
-        container_name = 'sysdig-probe-builder-{}'.format(dockerfile_tag)
+        #container_name = 'sysdig-probe-builder-{}'.format(dockerfile_tag)
+        container_name = ''
```
How worried are we about leftover containers after e.g. an interrupted build?

iurly replied:
good point. I guess the real threat is when you forcibly kill the main builder, in which case this won't really help I'm afraid (except at the next attempt).
We're ultimately using `docker --rm` which should clean up after itself, but you're right in that the following scenario will leave us with zombies:
- main builder spawned
- <max_jobs> probe builders spawned
- one or more of the above builders get stuck for whatever reason
- main builder gets killed by timeout or impatient user

at this point our best bet would be to have some external cleanup mechanism (either upon start or as a post step), based on either (a) the image name or (b) a prefix-based container name (e.g. container-name + random number).
I'd go for option (a), don't see much added value in (b).

gnosek commented on probe_builder/__init__.py:
```diff
@@ -124,8 +124,16 @@ def build(builder_image_prefix,
-    for release, target in kernel_dirs:
-        distro_builder.build_kernel(workspace, probe, distro.builder_distro, release, target)
+
+    parallelism = len(os.sched_getaffinity(0))
```
Maybe make it configurable, so (worst case) we can disable parallelism if we need to?

iurly replied:
This way it *sort of is* configurable. You just need to call the builder with the right value for `taskset`. :trollface: 
OK we already have a parameter for download parallelism, shall we add another one for build parallelism?
And yet another one for unpacker parallelism?
We're just running out of one-letter parameters but it all makes sense I guess.

gnosek replied:
Still, I'd rather have an option to run things sequentially if we run into issues, instead of reverting the PR.

iurly replied:
done, I chose `-j`

gnosek reviewed (dismissed):
Some minor nits but it's beautiful 🤩

iurly committed:
debug and print timing, cleanup and fix bug

address review comments and fix bug

deepskyblue86 commented on builder-entrypoint.sh:
```diff
@@ -60,11 +60,17 @@ build_bpf() {
 	fi
 }
 
+rm -rf /code/sysdig-rw
+cp -rf /code/sysdig-ro /code/sysdig-rw
```
:open_mouth: :heart_decoration:

gnosek reviewed (dismissed):
Nothing major, but I'd still prefer an option to do sequential builds if we run into issues

iurly committed:
parallel-builders: incorporate review comments

- make parallelism overridable through command line argument "-j"
- change wording about the chosen dockerfile
- re-add removed blank line

iurly reviewed (commented):
@gnosek stage's all yours now!
